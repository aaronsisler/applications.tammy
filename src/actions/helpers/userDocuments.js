export const setUserDocuments = (userDocuments) => ({
    type: 'SET_USER_DOCUMENTS',
    userDocuments,
});

// export const addUserDocument = (userDocument) => ({
//     type: 'ADD_USER_DOCUMENT',
//     userDocument,
// });
