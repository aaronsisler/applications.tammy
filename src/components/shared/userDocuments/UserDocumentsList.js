import React from 'react';
import PropTypes from 'prop-types';
import UserDocumentsListItem from './UserDocumentsListItem';

export default class UserDocumentsList extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (this.props.userDocuments.length === 0) {
            return (
                <div className="user_documents_list empty">
                    No available items
                </div>
            );
        }

        return (
            <div className="user_documents_list">
                {
                    this.props.userDocuments.map((userDocument, index) =>
                        <UserDocumentsListItem key={index} {...userDocument} />)
                }
            </div>
        );
    }
}

UserDocumentsList.propTypes = {
    userDocuments: PropTypes.array,
};
