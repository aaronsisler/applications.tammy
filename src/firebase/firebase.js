import * as firebase from 'firebase';

const config = {
    apiKey: "AIzaSyB1Usq_JhrJRmTKaHVjNMe5xzRmqT_U49E",
    authDomain: "tammy-in-hr.firebaseapp.com",
    databaseURL: "https://tammy-in-hr.firebaseio.com",
    projectId: "tammy-in-hr",
    storageBucket: "tammy-in-hr.appspot.com",
    messagingSenderId: "750914882179"
};

firebase.initializeApp(config);

const database = firebase.database();
const googleAuthProvider = new firebase.auth.GoogleAuthProvider();

export {
    database as default,
    firebase,
    googleAuthProvider,
};
