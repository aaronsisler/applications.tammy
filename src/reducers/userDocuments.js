const userDocumentsReducerDefaultState = [];

const userDocumentsReducer = (state = userDocumentsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_USER_DOCUMENTS':
            return action.userDocuments;
        default:
            return state;
    }
};

export default userDocumentsReducer;
