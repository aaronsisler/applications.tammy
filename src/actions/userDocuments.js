import database from '../firebase/firebase';
import { setUserDocuments } from 'Actions/helpers/userDocuments';

export const startSetUserDocuments = (userId) => (dispatch) =>
    database.ref(`user_documents/${userId}`).once('value').then((snapshot) => {
        const userDocuments = [];
        snapshot.forEach((childSnapshot) => {
            userDocuments.push({
                id: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        dispatch(setUserDocuments(userDocuments));
    })
