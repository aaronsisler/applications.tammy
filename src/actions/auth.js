import { firebase, googleAuthProvider } from '../firebase/firebase';
import { history } from 'Tools/history';

export const startLogin = (redirectUrl = '/dashboard') => () => {
    firebase.auth().signInWithPopup(googleAuthProvider)
        .then(() => {
            history.push(redirectUrl);
        })
}

export const startLogout = () => () => firebase.auth().signOut()
    .then(() => history.push('/logged_out'));
