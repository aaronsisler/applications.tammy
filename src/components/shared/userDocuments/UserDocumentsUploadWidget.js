import React from 'react';
import FileUploader from 'react-firebase-file-uploader';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Progress } from 'react-sweet-progress';
import { storage } from '../../../firebase/firebase';
import { startAddUserDocument } from 'Actions/userDocuments';


export class UserDocumentsUploadWidget extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            success: '',
            isUploading: false,
            progress: 0,
            error: ''
        };
        this.storageRef = `${this.props.userId}`;
    }

    handleUploadStart = () => this.setState({ isUploading: true, success: '', progress: 0 });

    handleProgress = (progress) => this.setState({ progress });

    handleUploadError = (error) => {
        this.setState({
            success: error,
            isUploading: false,
        });
    }

    handleUploadSuccess = (documentName) => {
        this.setState({ success: `${documentName} uploaded sucessfully`, progress: 100, isUploading: false });
        storage.ref(this.storageRef).child(documentName).getDownloadURL()
            .then(downloadURL => {
                this.props.startAddUserDocument(this.props.userId, { documentName, downloadURL })
            });
    };


    componentDidMount() {
    }

    render() {
        return (
            <div id="document_upload_widget">
                <div className="document_upload_content">
                    <div className="document_upload_content__loader">
                        <label className="button">
                            Upload a document
                            <FileUploader
                                accept="application/pdf, application/msword, .pdf, .docx, .doc"
                                hidden
                                name="document_upload"
                                storageRef={storage.ref(this.storageRef)}
                                onUploadStart={this.handleUploadStart}
                                onUploadError={this.handleUploadError}
                                onUploadSuccess={this.handleUploadSuccess}
                                onProgress={this.handleProgress}
                            />
                        </label>
                    </div>
                    {this.state.success &&
                        <div className="document_upload_content__success">
                            {this.state.success}
                        </div>
                    }
                    {this.state.isUploading &&
                        <div className="document_upload_content__progress">
                            <Progress
                                percent={this.state.progress}
                            />
                        </div>
                    }
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    userId: state.user && state.user.userId,
})

const mapDispatchToProps = (dispatch) => ({
    startAddUserDocument: (userId, userDocument) => dispatch(startAddUserDocument(userId, userDocument)),
})

export default connect(mapStateToProps, mapDispatchToProps)(UserDocumentsUploadWidget);

UserDocumentsUploadWidget.propTypes = {
    userId: PropTypes.string,
    startAddUserDocument: PropTypes.func.isRequired,
};
