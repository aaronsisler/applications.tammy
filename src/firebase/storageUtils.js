import { storage } from 'Firebase/firebase';

const retrieveDownloadUrl = (storagePath, documentName) =>
    storage.ref(storagePath).child(documentName).getDownloadURL()

export {
    retrieveDownloadUrl,
};
