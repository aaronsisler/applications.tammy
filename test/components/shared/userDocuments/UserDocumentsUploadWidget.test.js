import React from 'react';
import { shallow } from 'enzyme';
import { UserDocumentsUploadWidget } from 'Shared/userDocuments/UserDocumentsUploadWidget';
import { uid } from '../../../fixtures/auth';
import * as firebase from 'Firebase/firebase';
import { retrieveDownloadUrl } from 'Firebase/storageUtils';
jest.mock('Firebase/storageUtils')
retrieveDownloadUrl.mockReturnValue('mockDownloadUrl');

describe('UserDocumentsUploadWidget', () => {
    let wrapper;
    const startAddUserDocument = jest.fn();

    const buildWrapper = (userIdInput = uid) => {
        wrapper = shallow(
            <UserDocumentsUploadWidget
                startAddUserDocument={startAddUserDocument}
                userId={userIdInput}
            />
        );
    };

    beforeEach(() => {
        buildWrapper();
    });

    it('should render UserDocumentsUploadWidget correctly when success is populated', () => {
        expect(wrapper.setState({ success: 'Mocked Success' }));
        expect(wrapper).toMatchSnapshot();
    });

    it('should render UserDocumentsUploadWidget correctly when isUploading is true', () => {
        expect(wrapper.setState({ isUploading: true }));
        expect(wrapper).toMatchSnapshot();
    });

    it('should render UserDocumentsUploadWidget with the default state', () => {
        const defaultState = {
            success: '',
            isUploading: false,
            progress: 0,
            error: ''
        };

        expect(wrapper.state()).toEqual(defaultState);
        expect(wrapper).toMatchSnapshot();
    });

    describe('FirebaseFileUploader', () => {
        describe('storageRef', () => {
            it('should use the firebase storage ref', () => {
                expect(wrapper.find('FirebaseFileUploader').prop('storageRef')).toEqual(firebase.storage.ref(uid));
            });

            it('should have a path of userId', async () => {
                const refMock = jest.spyOn(firebase.storage, 'ref');

                await wrapper.find('FirebaseFileUploader').prop('onUploadStart')();

                expect(refMock).toHaveBeenLastCalledWith(uid);
            });
        });

        describe('onProgress() method', () => {
            it('should set progress', () => {
                const progress = 42;
                wrapper.find('FirebaseFileUploader').prop('onProgress')(progress);

                expect(wrapper.state('progress')).toBe(progress);
            });
        });

        describe('onUploadError() method', () => {
            it('should set success to error message', () => {
                const error = 'mockError';
                expect(wrapper.state('success')).toBe('');

                wrapper.find('FirebaseFileUploader').prop('onUploadError')(error);

                expect(wrapper.state('success')).toBe(error);
            });

            it('should set isUploading to false', () => {
                wrapper.setState({ isUploading: true });
                expect(wrapper.state('isUploading')).toBe(true);

                wrapper.find('FirebaseFileUploader').prop('onUploadError')();

                expect(wrapper.state('isUploading')).toBe(false);
            });
        });

        describe('onUploadStart() method', () => {
            it('should set isUploading to true', () => {
                expect(wrapper.state('isUploading')).toBe(false);

                wrapper.find('FirebaseFileUploader').prop('onUploadStart')();

                expect(wrapper.state('isUploading')).toBe(true);
            });

            it('should set success to empty', () => {
                wrapper.setState({ success: 'notEmpty' });
                expect(wrapper.state('success')).toBe('notEmpty');

                wrapper.find('FirebaseFileUploader').prop('onUploadStart')();

                expect(wrapper.state('success')).toBe('');
            });

            it('should set progress to zero', () => {
                wrapper.setState({ progress: 2 });
                expect(wrapper.state('progress')).toBe(2);

                wrapper.find('FirebaseFileUploader').prop('onUploadStart')();

                expect(wrapper.state('progress')).toBe(0);
            });
        });

        describe('onUploadSuccess() method', () => {
            const documentName = 'mockDocumentName';
            const message = `${documentName} uploaded sucessfully!`;
            const downloadUrl = 'mockDownloadUrl';

            it('should call retrieveDownloadUrl', async () => {
                await wrapper.find('FirebaseFileUploader').prop('onUploadSuccess')(documentName);

                expect(retrieveDownloadUrl).toHaveBeenLastCalledWith(uid, documentName);
            });

            it('should call startAddUserDocument', async () => {
                await wrapper.find('FirebaseFileUploader').prop('onUploadSuccess')(documentName);

                expect(startAddUserDocument).toHaveBeenLastCalledWith({ documentName, downloadUrl });
            });

            it('should set isUploading to false', async () => {
                wrapper.setState({ isUploading: true });
                expect(wrapper.state('isUploading')).toBe(true);

                await wrapper.find('FirebaseFileUploader').prop('onUploadSuccess')();

                expect(wrapper.state('isUploading')).toBe(false);
            });

            it('should set success to success message', async () => {
                await wrapper.find('FirebaseFileUploader').prop('onUploadSuccess')(documentName);

                expect(wrapper.state('success')).toBe(message);
            });

            it('should set progress to 100', async () => {
                await wrapper.find('FirebaseFileUploader').prop('onUploadSuccess')();

                expect(wrapper.state('progress')).toBe(100);
            });
        });
    });
});
