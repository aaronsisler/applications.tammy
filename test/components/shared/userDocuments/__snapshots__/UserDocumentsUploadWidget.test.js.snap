// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDocumentsUploadWidget should render UserDocumentsUploadWidget correctly when isUploading is true 1`] = `
<div
  id="user_documents_upload_widget"
>
  <div
    className="user_documents_upload_content"
  >
    <div
      className="user_documents_upload_content__loader"
    >
      <label
        className="button"
      >
        Upload a document
        <FirebaseFileUploader
          accept="application/pdf, application/msword, .pdf, .docx, .doc"
          hidden={true}
          name="user_documents_upload"
          onProgress={[Function]}
          onUploadError={[Function]}
          onUploadStart={[Function]}
          onUploadSuccess={[Function]}
          storageRef={
            Object {
              "child": [Function],
            }
          }
        />
      </label>
    </div>
    <div
      className="user_documents_upload_content__progress"
    >
      <t
        percent={0}
      />
    </div>
  </div>
</div>
`;

exports[`UserDocumentsUploadWidget should render UserDocumentsUploadWidget correctly when success is populated 1`] = `
<div
  id="user_documents_upload_widget"
>
  <div
    className="user_documents_upload_content"
  >
    <div
      className="user_documents_upload_content__loader"
    >
      <label
        className="button"
      >
        Upload a document
        <FirebaseFileUploader
          accept="application/pdf, application/msword, .pdf, .docx, .doc"
          hidden={true}
          name="user_documents_upload"
          onProgress={[Function]}
          onUploadError={[Function]}
          onUploadStart={[Function]}
          onUploadSuccess={[Function]}
          storageRef={
            Object {
              "child": [Function],
            }
          }
        />
      </label>
    </div>
    <div
      className="user_documents_upload_content__success"
    >
      Mocked Success
    </div>
  </div>
</div>
`;

exports[`UserDocumentsUploadWidget should render UserDocumentsUploadWidget with the default state 1`] = `
<div
  id="user_documents_upload_widget"
>
  <div
    className="user_documents_upload_content"
  >
    <div
      className="user_documents_upload_content__loader"
    >
      <label
        className="button"
      >
        Upload a document
        <FirebaseFileUploader
          accept="application/pdf, application/msword, .pdf, .docx, .doc"
          hidden={true}
          name="user_documents_upload"
          onProgress={[Function]}
          onUploadError={[Function]}
          onUploadStart={[Function]}
          onUploadSuccess={[Function]}
          storageRef={
            Object {
              "child": [Function],
            }
          }
        />
      </label>
    </div>
  </div>
</div>
`;
